File: frontend/src/services/ai/contextBuilder.ts
/**
 * Multi-File AI Context Manager
 * Builds intelligent context for AI requests
 * Based on how Bolt.new, Cursor, and Replit work
 */

import { ProjectFile } from '@/store/projectStore'

export interface AIContext {
  // Project Structure
  projectName: string
  projectType: 'react' | 'node' | 'nextjs' | 'vue' | 'python' | 'java' | 'unknown'
  fileTree: string // Text representation of folder structure

  // Selected Files (most relevant)
  selectedFiles: ContextFile[]

  // Project Metadata
  techStack: string[]
  dependencies?: Record<string, string>

  // Current State
  currentFile?: string
  cursorPosition?: { line: number, column: number }

  // Errors/Issues
  errors?: string[]
  warnings?: string[]

  // User Intent
  userGoal?: string
}

export interface ContextFile {
  path: string
  content: string
  language: string
  relevanceScore: number
  reason?: string // Why this file was included
}

export interface ContextBuildOptions {
  maxFiles?: number // Max files to include (default: 10)
  maxTokens?: number // Approximate token limit (default: 50000)
  includeTests?: boolean
  includeConfig?: boolean
}

/**
 * Build intelligent context from project files
 */
export function buildAIContext(
  userPrompt: string,
  project: {
    name: string
    files: ProjectFile[]
    selectedFile?: ProjectFile | null
  },
  options: ContextBuildOptions = {}
): AIContext {
  const {
    maxFiles = 10,
    maxTokens = 50000,
    includeTests = false,
    includeConfig = true
  } = options

  // Detect project type and tech stack
  const projectType = detectProjectType(project.files)
  const techStack = detectTechStack(project.files)

  // Build file tree representation
  const fileTree = buildFileTree(project.files)

  // Extract keywords from user prompt
  const keywords = extractKeywords(userPrompt)

  // Score and rank files by relevance
  const scoredFiles = project.files
    .map(file => ({
      file,
      score: calculateRelevance(file, keywords, project.selectedFile, {
        includeTests,
        includeConfig
      })
    }))
    .filter(({ score }) => score > 0)
    .sort((a, b) => b.score - a.score)
    .slice(0, maxFiles)

  // Convert to context files with token limiting
  const selectedFiles: ContextFile[] = []
  let totalTokens = 0

  for (const { file, score } of scoredFiles) {
    const fileTokens = estimateTokens(file.content || '')

    if (totalTokens + fileTokens > maxTokens) {
      break // Token limit reached
    }

    selectedFiles.push({
      path: file.path,
      content: file.content || '',
      language: file.language,
      relevanceScore: score,
      reason: getInclusionReason(file, project.selectedFile, keywords)
    })

    totalTokens += fileTokens
  }

  return {
    projectName: project.name,
    projectType,
    fileTree,
    selectedFiles,
    techStack,
    dependencies: extractDependencies(project.files),
    currentFile: project.selectedFile?.path,
    userGoal: userPrompt
  }
}

/**
 * Detect project type from files
 */
function detectProjectType(files: ProjectFile[]): AIContext['projectType'] {
  const hasFile = (pattern: string) =>
    files.some(f => f.path.match(new RegExp(pattern)))

  if (hasFile('package.json')) {
    if (hasFile('next.config')) return 'nextjs'
    if (hasFile('vite.config|App.tsx|App.jsx')) return 'react'
    if (hasFile('vue.config|App.vue')) return 'vue'
    return 'node'
  }

  if (hasFile('requirements.txt|setup.py|__init__.py')) return 'python'
  if (hasFile('pom.xml|build.gradle|.java$')) return 'java'

  return 'unknown'
}

/**
 * Detect tech stack
 */
function detectTechStack(files: ProjectFile[]): string[] {
  const stack: string[] = []
  const packageJson = files.find(f => f.path === 'package.json')

  if (packageJson?.content) {
    try {
      const pkg = JSON.parse(packageJson.content)
      const deps = { ...pkg.dependencies, ...pkg.devDependencies }

      // Detect frameworks
      if (deps.react) stack.push('React')
      if (deps.next) stack.push('Next.js')
      if (deps.vue) stack.push('Vue')
      if (deps.express) stack.push('Express')
      if (deps.nest || deps['@nestjs/core']) stack.push('NestJS')

      // Detect tools
      if (deps.typescript) stack.push('TypeScript')
      if (deps.tailwindcss) stack.push('Tailwind CSS')
      if (deps.vite) stack.push('Vite')
      if (deps.webpack) stack.push('Webpack')
    } catch (e) {
      // Invalid JSON
    }
  }

  // Detect from file extensions
  const hasExtension = (ext: string) =>
    files.some(f => f.path.endsWith(ext))

  if (hasExtension('.ts') || hasExtension('.tsx')) stack.push('TypeScript')
  if (hasExtension('.py')) stack.push('Python')
  if (hasExtension('.java')) stack.push('Java')

  return stack
}

/**
 * Calculate file relevance score
 */
function calculateRelevance(
  file: ProjectFile,
  keywords: string[],
  selectedFile: ProjectFile | null | undefined,
  options: { includeTests: boolean, includeConfig: boolean }
): number {
  let score = 0

  // Current/selected file gets highest priority
  if (selectedFile && file.path === selectedFile.path) {
    score += 100
  }

  // Keyword matching in filename
  const fileName = file.path.toLowerCase()
  for (const keyword of keywords) {
    if (fileName.includes(keyword.toLowerCase())) {
      score += 30
    }
  }

  // Keyword matching in content
  const content = (file.content || '').toLowerCase()
  for (const keyword of keywords) {
    if (content.includes(keyword.toLowerCase())) {
      score += 20
    }
  }

  // File type bonuses
  if (file.path.match(/\.(tsx?|jsx?)$/)) score += 15 // Source files
  if (file.path.includes('component')) score += 10
  if (file.path.includes('util') || file.path.includes('helper')) score += 8
  if (file.path.includes('type') || file.path.includes('interface')) score += 5

  // Penalties
  if (!options.includeTests && file.path.match(/\.(test|spec)\./)) score -= 50
  if (!options.includeConfig && file.path.match(/\.(config|rc)\./)) score -= 30
  if (file.path.includes('node_modules')) score = 0
  if (file.path.includes('.git')) score = 0
  if (file.path.match(/\.(lock|log)$/)) score = 0

  // Recently modified (if we had timestamps)
  // score += isRecent(file) ? 10 : 0

  // Import relationships (if we parse imports)
  // score += importsSelectedFile(file, selectedFile) ? 25 : 0

  return Math.max(0, score)
}

/**
 * Extract keywords from user prompt
 */
function extractKeywords(prompt: string): string[] {
  // Remove common words
  const stopWords = ['the', 'a', 'an', 'and', 'or', 'but', 'in', 'on', 'at', 'to', 'for', 'of', 'with', 'by', 'from', 'as', 'is', 'was', 'are', 'were', 'been', 'be', 'have', 'has', 'had', 'do', 'does', 'did', 'will', 'would', 'should', 'could', 'may', 'might', 'must', 'can', 'this', 'that', 'these', 'those']

  return prompt
    .toLowerCase()
    .replace(/[^\w\s]/g, ' ')
    .split(/\s+/)
    .filter(word => word.length > 2 && !stopWords.includes(word))
}

/**
 * Build file tree representation
 */
function buildFileTree(files: ProjectFile[]): string {
  const tree: Record<string, any> = {}

  for (const file of files) {
    const parts = file.path.split('/')
    let current = tree

    for (let i = 0; i < parts.length; i++) {
      const part = parts[i]
      if (i === parts.length - 1) {
        current[part] = file.type
      } else {
        if (!current[part]) current[part] = {}
        current = current[part]
      }
    }
  }

  return formatTree(tree, 0)
}

function formatTree(node: any, indent: number): string {
  let result = ''
  const entries = Object.entries(node)

  for (const [name, value] of entries) {
    const prefix = '  '.repeat(indent)
    const icon = typeof value === 'object' ? '📁' : '📄'
    result += `${prefix}${icon} ${name}\n`

    if (typeof value === 'object') {
      result += formatTree(value, indent + 1)
    }
  }

  return result
}

/**
 * Estimate token count
 */
function estimateTokens(text: string): number {
  // Rough estimate: 1 token ≈ 4 characters
  return Math.ceil(text.length / 4)
}

/**
 * Get reason for file inclusion
 */
function getInclusionReason(
  file: ProjectFile,
  selectedFile: ProjectFile | null | undefined,
  keywords: string[]
): string {
  if (selectedFile && file.path === selectedFile.path) {
    return 'Currently selected file'
  }

  const matchedKeywords = keywords.filter(k =>
    file.path.toLowerCase().includes(k) ||
    (file.content || '').toLowerCase().includes(k)
  )

  if (matchedKeywords.length > 0) {
    return `Matches keywords: ${matchedKeywords.join(', ')}`
  }

  if (file.path.match(/\.(tsx?|jsx?)$/)) {
    return 'Main source file'
  }

  return 'Related to project'
}

/**
 * Extract dependencies from package.json
 */
function extractDependencies(files: ProjectFile[]): Record<string, string> | undefined {
  const packageJson = files.find(f => f.path === 'package.json')

  if (packageJson?.content) {
    try {
      const pkg = JSON.parse(packageJson.content)
      return { ...pkg.dependencies, ...pkg.devDependencies }
    } catch (e) {
      return undefined
    }
  }

  return undefined
}

/**
 * Format context for AI prompt
 */
export function formatContextForAI(context: AIContext): string {
  let prompt = `# Project: ${context.projectName}\n\n`

  prompt += `## Project Type\n${context.projectType}\n\n`

  if (context.techStack.length > 0) {
    prompt += `## Tech Stack\n${context.techStack.join(', ')}\n\n`
  }

  prompt += `## File Structure\n\`\`\`\n${context.fileTree}\`\`\`\n\n`

  if (context.selectedFiles.length > 0) {
    prompt += `## Relevant Files\n\n`

    for (const file of context.selectedFiles) {
      prompt += `### ${file.path}\n`
      if (file.reason) {
        prompt += `*${file.reason}*\n\n`
      }
      prompt += `\`\`\`${file.language}\n${file.content}\n\`\`\`\n\n`
    }
  }

  if (context.currentFile) {
    prompt += `## Current File\n${context.currentFile}\n\n`
  }

  if (context.userGoal) {
    prompt += `## User Request\n${context.userGoal}\n\n`
  }

  return prompt
}

File: backend/app/modules/bolt/context_builder.py
"""
AI Context Builder for Bolt.new-style code generation
Builds intelligent context from project files for Claude API
"""

from typing import List, Dict, Optional, Set
import re
from dataclasses import dataclass
from app.core.logging_config import logger


@dataclass
class ContextFile:
    path: str
    content: str
    language: str
    relevance_score: float
    reason: str


@dataclass
class AIContext:
    project_name: str
    project_type: str
    file_tree: str
    selected_files: List[ContextFile]
    tech_stack: List[str]
    dependencies: Optional[Dict[str, str]]
    current_file: Optional[str]
    user_goal: str


class BoltContextBuilder:
    """Build intelligent context for AI requests"""

    # Stop words for keyword extraction
    STOP_WORDS = {
        'the', 'a', 'an', 'and', 'or', 'but', 'in', 'on', 'at', 'to', 'for',
        'of', 'with', 'by', 'from', 'as', 'is', 'was', 'are', 'were', 'been',
        'be', 'have', 'has', 'had', 'do', 'does', 'did', 'will', 'would',
        'should', 'could', 'may', 'might', 'must', 'can', 'this', 'that',
        'these', 'those'
    }

    def __init__(self):
        pass

    def build_context(
        self,
        user_prompt: str,
        files: List[Dict],
        project_name: str = "Project",
        selected_file_path: Optional[str] = None,
        max_files: int = 10,
        max_tokens: int = 50000
    ) -> AIContext:
        """
        Build intelligent context from project files

        Args:
            user_prompt: User's request
            files: List of file dicts with 'path', 'content', 'language'
            project_name: Project name
            selected_file_path: Currently selected file path
            max_files: Maximum files to include
            max_tokens: Approximate token limit

        Returns:
            AIContext object
        """
        # Detect project type and tech stack
        project_type = self._detect_project_type(files)
        tech_stack = self._detect_tech_stack(files)

        # Build file tree
        file_tree = self._build_file_tree(files)

        # Extract keywords from prompt
        keywords = self._extract_keywords(user_prompt)

        # Score and rank files
        scored_files = []
        for file in files:
            if file.get('type') == 'folder':
                continue

            score = self._calculate_relevance(
                file,
                keywords,
                selected_file_path
            )

            if score > 0:
                scored_files.append({
                    'file': file,
                    'score': score
                })

        # Sort by score and limit
        scored_files.sort(key=lambda x: x['score'], reverse=True)
        scored_files = scored_files[:max_files]

        # Convert to ContextFile objects with token limiting
        selected_files = []
        total_tokens = 0

        for item in scored_files:
            file = item['file']
            content = file.get('content', '')
            file_tokens = self._estimate_tokens(content)

            if total_tokens + file_tokens > max_tokens:
                break

            selected_files.append(ContextFile(
                path=file['path'],
                content=content,
                language=file.get('language', 'plaintext'),
                relevance_score=item['score'],
                reason=self._get_inclusion_reason(
                    file,
                    selected_file_path,
                    keywords
                )
            ))

            total_tokens += file_tokens

        return AIContext(
            project_name=project_name,
            project_type=project_type,
            file_tree=file_tree,
            selected_files=selected_files,
            tech_stack=tech_stack,
            dependencies=self._extract_dependencies(files),
            current_file=selected_file_path,
            user_goal=user_prompt
        )

    def format_for_claude(self, context: AIContext) -> str:
        """Format context for Claude API"""
        prompt = f"# Project: {context.project_name}\n\n"

        prompt += f"## Project Type\n{context.project_type}\n\n"

        if context.tech_stack:
            prompt += f"## Tech Stack\n{', '.join(context.tech_stack)}\n\n"

        prompt += f"## File Structure\n```\n{context.file_tree}```\n\n"

        if context.selected_files:
            prompt += "## Relevant Files\n\n"

            for file in context.selected_files:
                prompt += f"### {file.path}\n"
                if file.reason:
                    prompt += f"*{file.reason}*\n\n"
                prompt += f"```{file.language}\n{file.content}\n```\n\n"

        if context.current_file:
            prompt += f"## Current File\n{context.current_file}\n\n"

        prompt += f"## User Request\n{context.user_goal}\n\n"

        return prompt

    def _detect_project_type(self, files: List[Dict]) -> str:
        """Detect project type from files"""
        file_paths = [f['path'] for f in files]

        has_file = lambda pattern: any(re.search(pattern, path) for path in file_paths)

        if has_file(r'package\.json'):
            if has_file(r'next\.config'):
                return 'nextjs'
            if has_file(r'vite\.config|App\.(tsx|jsx)'):
                return 'react'
            if has_file(r'vue\.config|App\.vue'):
                return 'vue'
            return 'node'

        if has_file(r'requirements\.txt|setup\.py|__init__\.py'):
            return 'python'

        if has_file(r'pom\.xml|build\.gradle|\.java$'):
            return 'java'

        return 'unknown'

    def _detect_tech_stack(self, files: List[Dict]) -> List[str]:
        """Detect tech stack from files"""
        stack = []

        # Find package.json
        package_json = next(
            (f for f in files if f['path'] == 'package.json'),
            None
        )

        if package_json and package_json.get('content'):
            try:
                import json
                pkg = json.loads(package_json['content'])
                deps = {**pkg.get('dependencies', {}), **pkg.get('devDependencies', {})}

                if 'react' in deps:
                    stack.append('React')
                if 'next' in deps:
                    stack.append('Next.js')
                if 'vue' in deps:
                    stack.append('Vue')
                if 'express' in deps:
                    stack.append('Express')
                if 'typescript' in deps:
                    stack.append('TypeScript')
                if 'tailwindcss' in deps:
                    stack.append('Tailwind CSS')
                if 'vite' in deps:
                    stack.append('Vite')
            except (json.JSONDecodeError, IOError, OSError, KeyError) as e:
                logger.debug(f"Could not parse package.json for tech stack: {e}")

        # Detect from extensions
        has_ext = lambda ext: any(f['path'].endswith(ext) for f in files)

        if has_ext('.ts') or has_ext('.tsx'):
            if 'TypeScript' not in stack:
                stack.append('TypeScript')
        if has_ext('.py'):
            stack.append('Python')
        if has_ext('.java'):
            stack.append('Java')

        return stack

    def _calculate_relevance(
        self,
        file: Dict,
        keywords: Set[str],
        selected_file_path: Optional[str]
    ) -> float:
        """Calculate file relevance score"""
        score = 0.0

        # Current/selected file gets highest priority
        if selected_file_path and file['path'] == selected_file_path:
            score += 100

        # Keyword matching in filename
        file_name = file['path'].lower()
        for keyword in keywords:
            if keyword in file_name:
                score += 30

        # Keyword matching in content
        content = (file.get('content') or '').lower()
        for keyword in keywords:
            if keyword in content:
                score += 20

        # File type bonuses
        if re.search(r'\.(tsx?|jsx?)$', file['path']):
            score += 15  # Source files
        if 'component' in file_name:
            score += 10
        if 'util' in file_name or 'helper' in file_name:
            score += 8
        if 'type' in file_name or 'interface' in file_name:
            score += 5

        # Penalties
        if re.search(r'\.(test|spec)\.', file['path']):
            score -= 50
        if re.search(r'\.(config|rc)\.', file['path']):
            score -= 30
        if 'node_modules' in file['path']:
            score = 0
        if '.git' in file['path']:
            score = 0
        if re.search(r'\.(lock|log)$', file['path']):
            score = 0

        return max(0, score)

    def _extract_keywords(self, prompt: str) -> Set[str]:
        """Extract keywords from user prompt"""
        words = re.sub(r'[^\w\s]', ' ', prompt.lower()).split()
        keywords = {
            word for word in words
            if len(word) > 2 and word not in self.STOP_WORDS
        }
        return keywords

    def _build_file_tree(self, files: List[Dict]) -> str:
        """Build text representation of file tree"""
        tree = {}

        for file in files:
            parts = file['path'].split('/')
            current = tree

            for i, part in enumerate(parts):
                if i == len(parts) - 1:
                    current[part] = file.get('type', 'file')
                else:
                    if part not in current:
                        current[part] = {}
                    current = current[part]

        return self._format_tree(tree, 0)

    def _format_tree(self, node: Dict, indent: int) -> str:
        """Format tree structure"""
        result = ""
        entries = sorted(node.items())

        for name, value in entries:
            prefix = "  " * indent
            icon = "📁" if isinstance(value, dict) else "📄"
            result += f"{prefix}{icon} {name}\n"

            if isinstance(value, dict):
                result += self._format_tree(value, indent + 1)

        return result

    def _estimate_tokens(self, text: str) -> int:
        """Estimate token count (rough: 1 token ≈ 4 characters)"""
        return len(text) // 4

    def _get_inclusion_reason(
        self,
        file: Dict,
        selected_file_path: Optional[str],
        keywords: Set[str]
    ) -> str:
        """Get reason for file inclusion"""
        if selected_file_path and file['path'] == selected_file_path:
            return "Currently selected file"

        matched_keywords = [
            k for k in keywords
            if k in file['path'].lower() or k in (file.get('content') or '').lower()
        ]

        if matched_keywords:
            return f"Matches keywords: {', '.join(list(matched_keywords)[:3])}"

        if re.search(r'\.(tsx?|jsx?)$', file['path']):
            return "Main source file"

        return "Related to project"

    def _extract_dependencies(self, files: List[Dict]) -> Optional[Dict[str, str]]:
        """Extract dependencies from package.json"""
        package_json = next(
            (f for f in files if f['path'] == 'package.json'),
            None
        )

        if package_json and package_json.get('content'):
            try:
                import json
                pkg = json.loads(package_json['content'])
                return {
                    **pkg.get('dependencies', {}),
                    **pkg.get('devDependencies', {})
                }
            except (json.JSONDecodeError, TypeError, KeyError) as e:
                logger.debug(f"Could not parse package.json dependencies: {e}")

        return None


# Singleton instance
context_builder = BoltContextBuilder()
